<hr>

### Exercise 1.45

My favorite prime number is 8675309. Yep. Jenny’s phone number is prime! Write a script that verifies this fact.

- **Hint**: You only need to check divisors as large as the square root of 8675309 (why).



In [4]:
#### Solution 1.45 - Approach 1

# instead of checking only 8675309, we write a function that can be used for any number

import numpy as np

def is_prime(n):
    # Check if the number is less than 2 or even
    if n < 2 or n % 2 == 0:
        return False
    # Check for factors from 3 to the square root of n, skipping even numbers
    for i in range(3, int(np.sqrt(n)) + 1, 2):
        if n % i == 0:
            return False
    return True

number = 8675309
print(f'Approach 1: {number:,} is prime? {is_prime(number)}')

#### Approach 2 - A more optimized solution

# instead of checking only 8675309, we write a function that can be used for any number

import numpy as np

def is_prime(n):
    # Check if the number is less than 2 or even
    if n < 2:
        return False
    if n in (2, 3):
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    # Check for factors from 5 to the square root of n, skipping multiples of 2 and 3
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i + 2) == 0:
            return False
        i += 6
    return True

number = 8675309
print(f'Approach 2: {number:,} is prime? {is_prime(number)}')

Approach 1: 8,675,309 is prime? True
Approach 2: 8,675,309 is prime? True


<hr>

### Exercise 1.46

Write a function that accepts an integer and returns a binary variable:

- 0 = not prime,
- 1 = prime.

Next, write a script to find the sum of all of the prime numbers less than 1000.

**Hint:**
Remember that a prime number has exactly two divisors: 1 and itself. You only need to check divisors as large as the square root of $n$. Your script should probably be smart enough to avoid all of the non-prime even numbers.

(This problem is modified from [3])



In [5]:
#### Soution 1.46 - Approach 1

# in this approach we'll use the is_prime function from the previous solution which returns a boolean

limit = 1000
sum_primes = sum(x for x in range(limit) if is_prime(x))
print(f'Approach 1: Sum of primes below {limit} = {sum_primes}')

#### Approach 2 - We'll use a function that returns 0 or 1

def is_prime_zero_one(n):
    # returns 1 if n is prime, 0 otherwise
    return int(is_prime(n))

sum_primes = sum(is_prime_zero_one(x)*x for x in range(limit))
print(f'Approach 2: Sum of primes below {limit} = {sum_primes}')

Approach 1: Sum of primes below 1000 = 76127
Approach 2: Sum of primes below 1000 = 76127


<hr>

### Exercise 1.47

The sum of the squares of the first ten natural numbers is:

$$1^2 + 2^2 + \dots + 10^2 = 385$$

The square of the sum of the first ten natural numbers is:

$$(1 + 2 + \dots + 10)^2 = 55^2 = 3025$$

Hence, the difference between the square of the sum of the first ten natural numbers and the sum of the squares is:

$$3025 - 385 = 2640.$$

Write code to find the difference between the square of the sum of the first one hundred natural numbers and the sum of the squares. Your code needs to run error-free and output only the difference. 

(This problem is modified from [3])



In [3]:
#### Solution 1.47 - Approach 1 - using a simple loop

n = 100
total = 0
squares_total = 0

# Loop through the first n natural numbers
for i in range(1, n + 1):
    total += i  # Sum of the numbers
    squares_total += i ** 2  # Sum of the squares of the numbers

total_squared = total ** 2  # Square of the sum of the numbers
diff = total_squared - squares_total  # Difference between the square of the sum and the sum of the squares

# Print the results
print(f'Approach 1: For the first {n} natural numbers:')
print(f'The square of the sum is {total_squared:,}')
print(f'The sum of the squares is {squares_total:,}')
print(f'The difference of the square of the sum and sum of the squares is {diff:,}\n')

#### Approach 2 - more pythonic using list comprehensions

def square_sum_minus_sum_square(n):
    return sum(range(1, n+1))**2 - sum(x**2 for x in range(1, n+1))

n = 100
diff = square_sum_minus_sum_square(n)
print(f'Approach 2: For the first {n} natural numbers:')
print(f'The difference of the square of the sum and sum of the squares is {diff:,}')

Approach 1: For the first 100 natural numbers:
The square of the sum is 25,502,500
The sum of the squares is 338,350
The difference of the square of the sum and sum of the squares is 25,164,150

Approach 2: For the first 100 natural numbers:
The difference of the square of the sum and sum of the squares is 25,164,150


<hr>

### Exercise 1.48

The prime factors of $13195$ are $5, 7, 13,$ and $29$. Write code to find the largest prime factor of the number $600851475143$. Your code needs to run error-free and output only the largest prime factor.

(This problem is modified from [3])



In [28]:
#### Solution 1.48

# IDEA:  we can find the largest prime factor of a number by dividing it 
#        by its smallest prime factor until we reach 1

def largest_prime_factor(n):
    # Initialize the largest prime factor
    largest = 1
    # Check for factors of 2 and divide them out
    while n % 2 == 0:
        largest = 2
        n = n // 2
    # Check for odd factors from 3 to the square root of n and divide them out
    for i in range(3, int(np.sqrt(n)) + 1, 2):
        while n % i == 0:
            largest = i
            n = n // i
    # If n is greater than 2, it is prime
    if n > 2:
        largest = n
    return largest

print(f'The largest prime factor of 2025 is {largest_prime_factor(2025)}')
print(f'The largest prime factor of 13195 is {largest_prime_factor(13195)}')
print(f'The largest prime factor of 600851475143 is {largest_prime_factor(600851475143)}')

The largest prime factor of 2025 is 5
The largest prime factor of 13195 is 29
The largest prime factor of 600851475143 is 6857


### Exercise 1.51

Sometimes floating point arithmetic does not work like we would expect (and hope) as compared to by-hand mathematics. In each of the following problems, we have a mathematical problem that the computer gets wrong. Explain why the computer is getting these wrong.

a. Mathematically, we know that $ \sqrt{5}^2 $ should just give us 5 back. In Python, type `np.sqrt(5)**2 == 5`. What do you get and why do you get it?

In [44]:
#### Solution 1.51a

import numpy as np
print(f'np.sqrt(5)**2==5 is {np.sqrt(5)**2==5}.')
print(f'The absolute error is: {np.abs(np.sqrt(5)**2-5)}')

np.sqrt(5)**2==5 is False.
The absolute error is: 8.881784197001252e-16


$\sqrt{5}$ is an irrational number and has an infinite decimal (or binary) expansion.  When we represent it in the computer there will a small error since the finite binary expansion can't match exactly the infinite expansion of the true value.  In double precision, machine epsilon is $\epsilon = 2^{-52}$ and absolute round off error for a number $x$ is
$$ \text{Error} \approx |x| \cdot \epsilon = |x| \cdot 2^{-52}$$ 

For $\sqrt{5}$ the approximate absolute rounding error is

In [13]:
sqrt5_roundoff = np.sqrt(5)*2**(-52)
print( sqrt5_roundoff )

4.965068306494546e-16


The discrepancy between `np.sqrt(5)**2` and 5 is because of using this floating point approximation in our calculation.

  

b. Mathematically, we know that $ \left( \frac{1}{49} \right) \cdot 49 $ should just be 1. In Python, type `(1/49)*49 == 1`. What do you get and why do you get it?  


In [43]:
#### Solution 1.51b

import numpy as np
print(f'(1/49)*49 == 1 is {(1/49)*49==1}.')
print(f'The absolute error is: {np.abs( (1/49)*49-1 ) }')


(1/49)*49 == 1 is False.
The absolute error is: 1.1102230246251565e-16


The problem is the same.  $1/49$ will have a round off error in its floating point representation.


c. Mathematically, we know that $ e^{\ln(3)} $ should just give us 3 back. In Python, type `np.exp(np.log(3)) == 3`. What do you get and why do you get it?  


In [45]:
#### Solution 1.51c

import numpy as np
print(f'np.exp(np.log(3))==1 is {np.exp(np.log(3))==1}.')
print(f'The absolute error is: {np.abs(np.exp(np.log(3))-3)}')


np.exp(np.log(3))==1 is False.
The absolute error is: 4.440892098500626e-16


$\ln(3)$ is irrational leading to a round off error which affects the computation.


d. Create your own example of where Python gets something incorrect because of floating-point arithmetic.  

(This problem is modified from [4])  



In [16]:
#### Solution

(5**(1/3))**3 == 5

False

<hr>

### Exercise 1.52

In the 1999 movie *Office Space*, a character creates a program that takes fractions of cents that are truncated in a bank’s transactions and deposits them to his own account. This idea has been attempted in the past, and now banks look for this sort of thing. In this problem, you will build a simulation of the program to see how long it takes to become a millionaire.

**Assumptions:**
- Assume that you have access to 50,000 bank accounts.
- Assume that the account balances are uniformly distributed between $100 and $100,000.
- Assume that the annual interest rate on the accounts is 5% and the interest is compounded daily and added to the accounts, except that fractions of cents are truncated.
- Assume that your `illegal` account initially has a $0 balance.

**Your Tasks:**
a. Explain what the code below does:
```python
import numpy as np
accounts = 100 + (100000-100) * np.random.rand(50000, 1)
accounts = np.floor(100*accounts)/100
```



#### Solution

Creates 50,000 accounts with uniformly distributed initial values between $100 and $100,000.  

The last line rounds the value in each account down to the nearest penny.


b. By hand (no computer), write the mathematical steps necessary to increase the accounts by $(5/365)\%$ per day, truncate the accounts to the nearest penny, and add the truncated amount into an account titled “illegal.”


#### Solution

1.  compute the daily interest as $i = \text{account} \times 0.05/365$
2.  compute the interest rounded down to the nearest penny as $i_d = \text{floor}(i \times 100)/100$
3.  compute the residual interest as $i_r = i - i_d$
4.  add $i_d$ to the value of the account
5.  do this for all accounts and sum the values of $i_r$ and add to the "illegal" account

c. Write code to complete your plan from part (b).  


In [18]:
#### Solution to 1.52c

# to compute the total interest for a single day from all the accounts

accounts = 100 + (100000-100) * np.random.rand(50_000, 1)
rate = 0.05 / 365
illegal = 0
interest = accounts * rate
interest_rounded_down = np.floor(interest * 100) / 100
interest_to_steal = interest - interest_rounded_down
illegal = illegal + np.sum(interest_to_steal)
accounts = accounts + interest_rounded_down

print(f'Total stolen interest for the day: ${illegal:.2f}')

Total stolen interest for the day: $250.07



d. Using a `while` loop, iterate over your code until the illegal account has accumulated $1,000,000. How long does it take?

(This problem is modified from [4])  



In [19]:
#### Solution to 1.52d

accounts = 100 + (100000-100) * np.random.rand(50_000, 1)  # Generate random account balances between $100 and $100,000
rate = 0.05 / 365  # Calculate the daily interest rate
illegal = 0  # Initialize the illegal account balance
days = 0  # Initialize the number of days

while illegal < 1_000_000:  # Continue looping until the illegal account balance reaches $1,000,000
    days += 1  # Increment the number of days
    interest = accounts * rate  # Calculate the daily interest for each account
    interest_rounded_down = np.floor(interest * 100) / 100  # Round down the interest to the nearest penny
    accounts = accounts + interest_rounded_down  # Add the rounded interest to the account balances
    interest_to_steal = interest - interest_rounded_down  # Calculate the residual interest to be added to the illegal account
    illegal = illegal + np.sum(interest_to_steal)  # Add the residual interest to the illegal account balance

print(f'It will take {days} days to steal $1,000,000.')
print(f'Or about {days/365:.2f} years.')

It will take 3998 days to steal $1,000,000.
Or about 10.95 years.


<hr>

### Exercise 1.53

In the 1991 Gulf War, the Patriot missile defense system failed due to roundoff error. The troubles stemmed from a computer that performed the tracking calculations with an internal clock whose integer values in tenths of a second were converted to seconds by multiplying by a 24-bit binary approximation to $0.1$:

$$
0.1_{10} \approx 0.00011001100110011001100_2
$$

- **a.** Convert the binary number above to a fraction by hand (common denominators would be helpful).


In [47]:
2**21

2097152

#### Solution

$$
\begin{align}
0.00011001100110011001100_2 &= 2^{-4} + 2^{-5} + 2^{-8} + 2^{-9} + 2^{-12} + 2^{-13} + 2^{-16}+2^{-17} + 2^{-20} + 2^{-21} \\
&= 2^{-21} ( 2^{17} + 2^{16}+2^{13} + 2^{12} + 2^9 + 2^8 + 2^5 + 2^4 + 2^1 + 2^0) \\
&= \frac{209,715}{2,097,152}
\end{align}
$$

- **b.** The approximation of $0.1$ given above is clearly not equal to $0.1$. What is the absolute error in this value?


In [49]:
# 1.53b 

abs_err = np.abs(209_715 / 2_097_152 - 1/10)
print(f'The approximate absolute error is: {abs_err}')

The approximate absolute error is: 9.536743164617612e-08


- **c.** What is the time error, in seconds, after 100 hours of operation?


In [53]:
# 1.53c Solution

# error per second
error_per_second = abs_err * 10

# Compute 100 hours in seconds
hours = 100
seconds = hours * 60 * 60

# total time error in seconds
time_error = error_per_second * seconds

print(f'The total time error in 100 hours is {time_error:.2f} seconds.')


The total time error in 100 hours is 0.34 seconds.


- **d.** During the 1991 war, a Scud missile traveled at approximately Mach 5 (3750 mph). Find the distance that the Scud missile would travel during the time error computed in (c).



In [55]:
# 1.53d Solution

speed_mps = 3750 * 1/3600 # multiply by 1/3600 to convert from mph to mps
distance = speed_mps * time_error

print(f'The missile travels about {distance:.2f} miles off course.')

The missile travels about 0.36 miles off course.
